hibernate-core/src/main/java/org/hibernate/query/sqm/internal/QuerySqmImpl.java

```diff
@@ -1008,9 +1007,32 @@ public R getSingleResult() {
 	// Named query externalization
 
 	@Override
-	public NamedHqlQueryMemento toMemento(String name) {
+	public NamedQueryMemento toMemento(String name) {
 		if ( CRITERIA_HQL_STRING.equals( getQueryString() ) ) {
-			throw new UnsupportedOperationException( "Criteria-based Query cannot be saved as a named query" );
+			final SqmStatement sqmStatement ;
+			if ( getSession().isJpaCriteriaCopyComplianceEnabled() ) {
```

Comment:
Since we already copy in the constructor of this class if the compliance is enabled, my suggestion was more that we copy this if compliance is disabled, to be safe. 

```suggestion
			if ( !getSession().isJpaCriteriaCopyComplianceEnabled() ) {
```